From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 10 Sep 2024 21:35:48 +0300
Subject: [PATCH] Data-driven items

Displaying items to the client is done via Item Modifiers API in a later patch.

Vanilla /give command won't show new items in auto complete.

diff --git a/src/main/java/io/papermc/paper/registry/event/RegistryEvents.java b/src/main/java/io/papermc/paper/registry/event/RegistryEvents.java
index 40deffbd0930508bb04e9aedfd62ad2144855198..0e95fb15bc11aae249dd2fab77cb7785cbff87e5 100644
--- a/src/main/java/io/papermc/paper/registry/event/RegistryEvents.java
+++ b/src/main/java/io/papermc/paper/registry/event/RegistryEvents.java
@@ -23,6 +23,7 @@ public final class RegistryEvents {
     public static final RegistryEventProvider<GameEvent, GameEventRegistryEntry.Builder> GAME_EVENT = create(RegistryKey.GAME_EVENT);
     public static final RegistryEventProvider<Enchantment, EnchantmentRegistryEntry.Builder> ENCHANTMENT = create(RegistryKey.ENCHANTMENT);
     public static final RegistryEventProvider<Art, PaintingVariantRegistryEntry.Builder> PAINTING_VARIANT = create(RegistryKey.PAINTING_VARIANT);
+    public static final RegistryEventProvider<org.bukkit.inventory.ItemType, me.sosedik.kiterino.registry.data.ItemRegistryEntity.Builder> ITEM = create(RegistryKey.ITEM); // Kiterino - Data-driven items
 
     private RegistryEvents() {
     }
diff --git a/src/main/java/me/sosedik/kiterino/registry/data/ItemRegistryEntity.java b/src/main/java/me/sosedik/kiterino/registry/data/ItemRegistryEntity.java
new file mode 100644
index 0000000000000000000000000000000000000000..fd8e781980e3bd585269c0816743127c2de7db2b
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/registry/data/ItemRegistryEntity.java
@@ -0,0 +1,101 @@
+package me.sosedik.kiterino.registry.data;
+
+import io.papermc.paper.datacomponent.DataComponentType;
+import io.papermc.paper.registry.RegistryBuilder;
+import org.bukkit.inventory.ItemType;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.Range;
+
+/**
+ * A data-centric version-specific registry entry for the {@link ItemType} type.
+ */
+@ApiStatus.Experimental
+@ApiStatus.NonExtendable
+public interface ItemRegistryEntity {
+
+    /**
+     * Gets the item's component data
+     *
+     * @return component data or null
+     */
+    boolean hasData(@NonNull DataComponentType type);
+
+    /**
+     * Gets the item's component data
+     *
+     * @return component data or null
+     * @param <T> component data type
+     */
+    <T> @Nullable T getData(DataComponentType.@NonNull Valued<T> type);
+
+    /**
+     * Provides the item's compost chance
+     *
+     * @return the item's compost chance
+     */
+    @Nullable Float compostChance();
+
+    /**
+     * Provides the item's nms implementation
+     *
+     * @return the item's nms implementation
+     */
+    @Nullable Object nmsItem();
+
+    /**
+     * Constructs item properties with item id already set
+     *
+     * @return item properties
+     */
+    @NonNull Object constructItemProperties();
+
+    /**
+     * A mutable builder for the {@link ItemRegistryEntity} plugins may change in applicable registry events.
+     */
+    @ApiStatus.Experimental
+    @ApiStatus.NonExtendable
+    interface Builder extends ItemRegistryEntity, RegistryBuilder<ItemType> {
+
+        /**
+         * Sets item's component data
+         *
+         * @param type component type
+         * @param value component value
+         * @param <T> component data type
+         */
+        @Contract(value = "_ -> this", mutates = "this")
+        <T> @NonNull Builder setData(DataComponentType.@NonNull Valued<T> type, @Nullable T value);
+
+        /**
+         * Sets item's component data
+         *
+         * @param type component type
+         * @param <T> component data type
+         */
+        @Contract(value = "_ -> this", mutates = "this")
+        <T> @NonNull Builder setData(DataComponentType.@NonNull NonValued type);
+
+        /**
+         * Configures the item's compost chance
+         *
+         * @param levelIncreaseChance the item's compost chance
+         * @return this builder
+         */
+        @Contract(value = "_ -> this", mutates = "this")
+        @NonNull Builder compostChance(@Range(from = 0L, to = 1L) @Nullable Float levelIncreaseChance);
+
+        /**
+         * Configures the item's nms implementation
+         *
+         * @param nmsItem the item's nms implementation
+         * @return this builder
+         */
+        @Contract(value = "_ -> this", mutates = "this")
+        @NonNull Builder nmsItem(@Nullable Object nmsItem);
+
+    }
+
+}
